### DISCUSSION TITLE
How to replace textlines in text under image double-layer pdf files?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#2657] [Q&A]
How to replace textlines in text under image double-layer pdf files?

**Comment by JorjMcKie** [comment]
Please use the `Discussions` tab for questions.

I don't understand what you mean: OCRed text under the scanned image? Something around Optional Content? ...?

**Comment by nissansz** [comment]
Yes. Abbyy or acrobat result exists on pdf.
But I want to replace content in each line.

**Reply by JorjMcKie** [reply]
What about using redaction annotations? You can prevent touching the image above by `Page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_NONE)`.

**Comment by nissansz** [comment]
Any full code sample?
Replace 'test' with 'ABC'
[Table1.pdf](https://github.com/pymupdf/PyMuPDF/files/12567763/Table1.pdf)

How to identify shape types: like paragraph, table, image, etc.?
And how to convert table to html?

**Reply by JorjMcKie** [reply]
Example:
```python
In [1]: import fitz
In [2]: doc=fitz.open("Table1.pdf")
In [3]: page=doc[0]
In [4]: r=page.search_for("we are3")[0]
In [5]: annot = page.add_redact_annot(r, "whatever")
In [6]: page.apply_redactions()
Out[6]: True
In [7]: doc.ez_save("redacted.pdf")
```
Gives you this:
![grafik](https://github.com/pymupdf/PyMuPDF/assets/8290722/249bfbb3-b1a0-4d6e-9b1b-37d60db3f356)

**Reply by JorjMcKie** [reply]
> How to identify shape types: like paragraph, table, image, etc.?

There is `page.get_bboxlog()` which returns all bboxes including their types.

**Reply by JorjMcKie** [reply]
>  And how to convert table to html?

There is no solution except coverting the whole page to HTML (`page.get_text("html")`).

**Reply by nissansz** [reply]
(page.get_text("html")

reuslt does not give table structure: <table><tr>

**Reply by JorjMcKie** [reply]
correct

**Reply by nissansz** [reply]
Any wayt to get table structure for correct html code? table, tr, td, etc.

**Reply by JorjMcKie** [reply]
No, there is no way.

**Comment by nissansz** [comment]
Instead of search page, I want to go through all lines, then replace whole line directly

  for i in range(0,doc.page_count):
        page = doc[i]
        j = 0
        for block in page.get_text("dict", flags=fitz.TEXTFLAGS_TEXT)["blocks"]:
            for line in block["lines"]:
                j=j+1
                bbox = line["bbox"]  # the line bbox
                text = ''.join([span["text"] for span in line["spans"]])  # text in line

**Reply by JorjMcKie** [reply]
Yes that also works - just use the bbox of the line for redactions.
